src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionJsonTypeInfoOfT.cs

```diff
@@ -37,11 +32,17 @@ internal ReflectionJsonTypeInfo(JsonConverter converter, JsonSerializerOptions o
             converter.ConfigureJsonTypeInfoUsingReflection(this, options);
         }
 
-        [RequiresUnreferencedCode(JsonSerializer.SerializationUnreferencedCodeMessage)]
-        [RequiresDynamicCode(JsonSerializer.SerializationRequiresDynamicCodeMessage)]
-        private void AddPropertiesAndParametersUsingReflection()
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2075:RequiresUnreferencedCode",
+            Justification = "The ctor is marked RequiresUnreferencedCode.")]
+        internal override void LateAddProperties()
         {
-            Debug.Assert(Converter.ConverterStrategy == ConverterStrategy.Object);
+            Debug.Assert(!IsConfigured);
+            Debug.Assert(PropertyCache is null);
+
+            if (Kind != JsonTypeInfoKind.Object)
+            {
```

Comment:
It seems that this can only be `true` when `JsonTypeInfo.Properties` is called. Should there be a check earlier in the call stack (say in the `Properties` getter itself or `PopulatePropertyList()`) that returns early with an empty list?

The check would then be reversed back to a `Debug.Assert`.

Reply:
I gave it a try, but it seems like the assert is failing. Given that we're backporting this, I'd rather we didn't make changes to control flow right now.

Reply:
FWIW I took the branching from the sourcegen implementation, so better to be consistent.